src/compiler/checker.ts

```diff
@@ -13833,10 +13833,22 @@ namespace ts {
                         // Infer to the simplified version of an indexed access, if possible, to (hopefully) expose more bare type parameters to the inference engine
                         const simplified = getSimplifiedType(target);
                         if (simplified !== target) {
-                            const key = source.id + "," + simplified.id;
-                            if (!visited || !visited.get(key)) {
-                                (visited || (visited = createMap<boolean>())).set(key, true);
-                                inferFromTypes(source, simplified);
+                            inferFromTypesOnce(source, simplified);
+                        }
+                        else if (target.flags & TypeFlags.IndexedAccess) {
+                            const indexType = getSimplifiedType((target as IndexedAccessType).indexType);
+                            // Generally simplifications of instantiable indexes are avoided to keep relationship checking correct, however if our target is an access, we can consider
+                            // that key of that access to be "instantiated", since we're looking to find the infernce goal in any way we can.
+                            if (indexType.flags & TypeFlags.Instantiable) {
+                                const objectType = getSimplifiedType((target as IndexedAccessType).objectType);
+                                // (T | U)[K] -> T[K] | U[K]
+                                if (objectType.flags & TypeFlags.Union) {

```

Comment:
pls test

Reply:
I've rewritten it to use functionality we use elsewhere to make synchronize it and make it evident that the union case is for consistency (as now it's just a call to the underlying distribution method). It's not actually possible to (meaningfully) take the union branch, since a reverse mapped type with a union for a key type would be an intersection of keys, which we do not currently produce inferences for (unlike unions, which we now do).